### PR TITLE
JSON API based on simdjson

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Functions to extract JSON values / perform pattern matching on JSON; Based on [simdjson](https://github.com/lemire/simdjson).

### CODE REVIEW

**Comment by hczhcz** [comment]
I have seen @mfridental 's pull request #4686 . Nice job!
Here is our solution to the limitation of exist JSON functions. Hopefully it can bring some help in API design, implementation, etc, no matter if it will be merged.

**Comment by mfridental** [comment]
Nice job too. I was afraid introducing a new dependency because I don't possess a good understanding of the build system. I also didn't discuss the new API in the group, because I wasn't sure whether I am able to implement it properly. 

About your solution: whether it can be considered as the beginning of the new DataType Json? After parsing incoming JSON, it could store it it some more efficient and searchable format than ColumnString? Is it possible to store JSON in a way that the column store nature of Clickhouse remains so that working with it is not slower than with normal columns?

**Comment by hczhcz** [comment]
@mfridental Good questions!

This PR is not intended to introduce a JSON datatype. Instead, it extracts structured data from JSON strings.

The scenario we meet in production is handling tables imported from somewhere else that contains JSON strings. Generally, we extracts the data from JSON strings. After that, we would either compose a big query (which returns all the information we need) or materialize the data for future queries.

**Comment by zhang2014** [comment]
@mfridental I am trying to finish this work : )

**Comment by vitlibar** [comment]
@hczhcz Can you fix build? I'd like to merge your PR but all checks should pass.

**Comment by hczhcz** [comment]
@vitlibar Sure. I will try to fix it.

**Comment by vitlibar** [comment]
simdjson shows good performance:

```
SELECT any(jsonExtractInt(json, 'a')) from (select '{"a":2,"b":35.4,"c":[1,2,{"e":"f\\n"}], "d":"qq"}' as json from system.numbers limit 200000000)
1 rows in set. Elapsed: 1.377 sec. Processed 200.02 million rows, 1.60 GB (145.20 million rows/s., 1.16 GB/s.)
```

```
SELECT any(jsonExtractString(json, 'c',3,'e')) from (select '{"a":2,"b":35.4,"c":[1,2,{"e":"f\\n"}], "d":"qq"}' as json from system.numbers limit 200000000)
1 rows in set. Elapsed: 1.774 sec. Processed 200.02 million rows, 1.60 GB (112.72 million rows/s., 901.79 MB/s.)
```

We can compare these results with summing numbers:
```
SELECT sum(number) from (select number from system.numbers limit 200000000)
1 rows in set. Elapsed: 0.466 sec. Processed 200.02 million rows, 1.60 GB (429.43 million rows/s., 3.44 GB/s.) 
```

File: dbms/src/Functions/FunctionsJSON.h [hunk 1]
```diff
@@ -14,7 +14,9 @@
     #pragma clang diagnostic ignored "-Wold-style-cast"
 #endif
 
-#include <simdjson/jsonparser.h>
+// use singleheader version to avoid extra objects from simdjson
```
Comment (vitlibar) [inline dbms/src/Functions/FunctionsJSON.h]: Why do you want to use the singleheader version?

File: dbms/src/Functions/FunctionsJSON.cpp [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+class JSONTypeImpl : public JSONNullableImplBase<DataTypeUInt8>
+{
+public:
+    static constexpr auto name{"jsonType"};
+
+    static Field getValue(ParsedJson::iterator & pjh)
+    {
+        // types: [{"sltfn
+        return pjh.get_type();
```
Comment (vitlibar) [inline dbms/src/Functions/FunctionsJSON.cpp]: Seems it's better to return string, not a character code.

File: dbms/src/Functions/FunctionsJSON.cpp [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+public:
+    static constexpr auto name{"jsonHas"};
+
+    static Field getValue(ParsedJson::iterator &) { return {1}; }
+};
+
+class JSONLengthImpl : public JSONNullableImplBase<DataTypeUInt64>
+{
+public:
+    static constexpr auto name{"jsonLength"};
```
Comment (vitlibar) [inline dbms/src/Functions/FunctionsJSON.cpp]: It's not easy to understand what `jsonLength()` is, and why `jsonLength('{"a":2,"b":35.4,"c":[1,2,{"e":"f\\n"}], "d":"qq"}')` is 8 and not 4.

Comment (vitlibar) [reply dbms/src/Functions/FunctionsJSON.cpp]: Could you please write some documentation and put it into https://github.com/yandex/ClickHouse/blob/master/docs/en/query_language/functions/json_functions.md ?

Comment (hczhcz) [reply dbms/src/Functions/FunctionsJSON.cpp]: Simdjson counts both keys and values when iterate through an object.

How about accessing keys only? So {"a":1,"b":2} will have `length = 2`, `[1] = "a"`, and `[2] = "b"`.

Comment (hczhcz) [reply dbms/src/Functions/FunctionsJSON.cpp]: @vitlibar Ok. I will push some updates tomorrow (GMT+8).

Comment (vitlibar) [reply dbms/src/Functions/FunctionsJSON.cpp]: Probably iterating this way would be slower. I suppose with documentation the current approach is ok.

File: dbms/src/Functions/CMakeLists.txt [hunk 4]
```diff
@@ -69,3 +69,8 @@ if (USE_HYPERSCAN)
     target_link_libraries (clickhouse_functions PRIVATE ${HYPERSCAN_LIBRARY})
     target_include_directories (clickhouse_functions SYSTEM PRIVATE ${HYPERSCAN_INCLUDE_DIR})
 endif ()
+
+if (HAVE_AVX2)
```
Comment (vitlibar) [inline dbms/src/Functions/CMakeLists.txt]: Did you define `HAVE_AVX2` somewhere?

File: dbms/src/Functions/FunctionsJSON.cpp [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            case 'l':
+                return "Int64";
+            case 'd':
+                return "Float64";
+            case 't':
+                return "Bool";
+            case 'f':
+                return "Bool";
+            case 'n':
+                return "Nothing";
```
Comment (vitlibar) [inline dbms/src/Functions/FunctionsJSON.cpp]: Why do you name it "Nothing"? It's `null` literal, isn't it?

File: dbms/src/Functions/CMakeLists.txt (hunk 4 above)
Comment (hczhcz) [reply dbms/src/Functions/CMakeLists.txt]: It is generated by cmake. It still have some trouble here.

File: dbms/src/Functions/FunctionsJSON.cpp (hunk 5 above)
Comment (hczhcz) [reply dbms/src/Functions/FunctionsJSON.cpp]: Make sense

**Comment by vitlibar** [comment]
Merged in https://github.com/yandex/ClickHouse/pull/5124

**Pull request closed without merging**